### PR TITLE
Blaze: Implement promote with blaze web view flow

### PR DESCRIPTION
**Pull request by zwarm**
Parent #18001 

This PR launches the Blaze flow inside of a webview.

Post List Light Header | Post List Dark Header
-- | --
Alt desc | Alt desc

Created the following to manage the Blaze create campaign webview flow. 
Includes: 
- Created `BlazeWebViewFragment`
- Created `BlazeWebViewViewModel`
- Created `BlazeWebClient`
- Created `BlazeActionEvent` for communication from the BlazeWebViewViewModel back to the BlazeWebViewFragment. For situation for finishing activity on error or flow ended
- Created and added the `BlazeWebViewHeaderUiState` and `BlazeWebViewContentUiState` states to the BlazeUiStates class. I toyed with using "model" vs. "UiState" suffix and am willing to rename as many times as necessary
- Do not allow the user to "back" out of the flow, they must use the cancel button
- "Blaze" is aligned to the left in the top header
- Tracking for blaze flow: started, canceled, error, and completed
 
Refactored: 
BlazeUiState:
 - Got rid of non-native uiStates to reflect the fact that we do not manage the "webView" flow like we do for a native flow.
 - Added "WebViewScreen" as a state
 - Added "page" as state (this was added as an entry point in the tech spec last week)
 - Added a getSource() to BlazeViewModel so I can grab it and pass along to BlazeWebViewViewModel )
 - Any errors encountered will push finishing the activity; push the user back to their entry point

BlazeParentActivity
- Launch the web view flow when the state changes to WebViewScreen 

BlazeViewModel
- Added the update for uiState onPromoteWithBlazeClicked (yes, I know this is a button click will be updated, but I wanted to make sure the showNextScreen worked)
- Adjusted the `showNextScreen` fun
- Added getSource() so I can grab the value in the BlazeWebView flow 

**Outstanding items - These will be handled in separate PRs**
- Page entry points handling
- Unit tests
- Compose look/feel of the header with regards to text placement/color (disabled doesn't look disabled yet) and webview in dark mode. This will go in a separate PR so Shaun can review as well.

**Known Issues**
- Disabling of the Cancel header action while payment is processing. This is being addressed on the server side.
- Tapping a "reader" external link launches the app reader, not in an external browser. Not sure how to approach this one yet.

**To test:**
**Setup**
- Install and launch the app
- Navigate to Me -> App Settings -> Privacy Setting
- Enable collection information
- Navigate to Me -> App Settings -> Debug Setting
- Enable Blaze
- Restart the app

**Test Dashboard Entry Point**
- Navigate to My SIte tab
- Tap on the Promote With Blaze card
- Tap on the Blaze a Post Now button within the overlay
- ✅ Verify you are launched into the Blaze flow post lists
- ✅ Verify logs contain:  🔵 Tracked: blaze_flow_started, Properties: {"source":"dashboard_card"}
- Tap the cancel button in the header
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"posts_list"}
- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Tap the cancel button in the header Appearance view
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"step_1"}

- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Fill out the Appearance information and tap Next
- Tap the cancel button in the header Audience view
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"step_2"}

- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Fill out the Appearance information and tap Next
- Fill out the Audience information and tap Next
- Tap the cancel button in the header on the Payment view
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"step_3"}

- Enter the flow again: Tap on the Promote With Blaze card > Tap on the Blaze a Post Now button within the overlay
- Select a post from the list and tap Promote
- Fill out the Appearance information and tap Next (Change the title and snippet to DO NOT APPROVE - TESTING)
- Fill out the Audience information and tap Next
- Fill out the Payment information and tap Save & Submit
- Note that the cancel button will still be visible and clickable, please do not click it. Need server change.
- Wait for the process to complete
-  ✅ Verify the header action button changed to "done"
- DO NOT TAP DONE YET (unless you want to go back through the process again ;) )
- Tap the TermsConditions link
-  ✅ Verify an external browser is launched
- Navigate back to the app
- Tap the View Campaigns button
- Tap the Header "Done" button
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_completed, Properties: {"source":"dashboard_card"}

**Test Post Entry**
- Navigate to My SIte tab
- Tap on the Posts menu item or ribbon action
- Tap on More > Promote with Blaze
- ✅ Verify you are launched into the Blaze flow at the Appearance page for the post you selected
- ✅ Verify logs contain:  🔵 Tracked: blaze_flow_started, Properties: {"source":"posts_list"}
- Tap the cancel button in the header
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"posts_list","current_step":"step_1"}
* this is a streamlined test, as the only thing that changes from the first set of tests is the source. Feel free to go through the complete flow if you want to verify this.

## Regression Notes
1. Potential unintended areas of impact
Blaze flow does not work properly

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Separate PR

PR submission checklist:
- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeActionEvent.kt
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeActionEvent.kt:4]: Hey @zwarm 👋🏼 

🗒️  I have added some logic for finishing the activity as well on close/back click of the overlay, I can update my logic to use this event after these changes land in the `trunk`. So you might see that logic in [PR](https://github.com/wordpress-mobile/WordPress-Android/pull/18018), Please feel free to ignore that during the review. 

Having a separate event for Action event 👏🏼 . Nice.

Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeActionEvent.kt:4]: @AjeshRPai Thanks for the heads up. After Done is clicked, I am passing back the FinishActivity which is being handled in the BlazeWebViewFragment. 🎈

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazeoverlay/BlazeOverlayFragment.kt
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazeoverlay/BlazeOverlayFragment.kt:20]: ❤️ 👏🏼   .Thanks for removing this line. I didn't realise that I had pushed this.

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeUiState.kt
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeUiState.kt]: 🔍 Nitpick: the `done` word is in lower case. I think we should show **Done** , right now its **done**. 

screenshot_nightMode_done_button_

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeFeatureUtils.kt
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeFeatureUtils.kt]: ❓ Any particular reason why this logic is extracted to the utils class? It looks like the logic is related to the UI more. 

💡 I would suggest moving this to the `BlazeWebViewViewModel`. But leave it to you if you want to keep it here.

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt:130]: 🔍 Nitpick: Can we move this under `TopAppBar`?. Just want to arrange the views according to the way they are rendered/called

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt]: 💡 Can we move this `when`check to the ViewModel. I think exposing a click function in ViewModel and then checking which state in `BlazeWebViewViewModel` would be better. The rule I follow: the lesser the functions exposed from ViewModel, the better, also less logic in UI.

**Review by AjeshRPai — APPROVED** [review]
Hey @zwarm 👋🏼 

I have tested the changes in this PR according to the Instructions you left in the PR. Everything looks great to me. I created a blaze campaign with the changes in this PR. 🚀 😄 . You owe me 5$. 😄 . 

I have left some 🔍 Nitpicks for you to check and make changes if necessary. No changes are blockers. But if you want to do it in the next PR which fixes the Known issues you have pointed out in PR description, that sounds good to me. I am approving but not merging so you can look at the Nitpicks and Suggestions.

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeUiState.kt
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeUiState.kt]: Thanks @AjeshRPai - this was already on my list for the UI improvements PR, but thanks for making the issue visible to others. Fixed in commit [5d6e5e5](https://github.com/wordpress-mobile/WordPress-Android/pull/18013/commits/5d6e5e5750cfd5acb0dfba04315e43c25cc86f26)

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeFeatureUtils.kt
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/blaze/BlazeFeatureUtils.kt]: @AjeshRPai - I had planned on using it elsewhere, but it makes sense to move it over to the `BlazeWebViewViewModel`. Thanks for the suggestion. In commit [215b633](https://github.com/wordpress-mobile/WordPress-Android/pull/18013/commits/215b633ac9387b101d2ad225472c779e144e16ae)

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt:130]: Absolutely. Good idea. In commit [9ea859b](https://github.com/wordpress-mobile/WordPress-Android/pull/18013/commits/9ea859b87670075c9f997cfd7a12b9420096a4cb)

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/blaze/ui/blazewebview/BlazeWebViewFragment.kt]: Love this idea, thanks for the suggestion. ❤️  Ideally the click handler should be in the state, but I couldn't get it working properly, so this will do for now. I will revisit in the future when I have more time.
In commit [b79fa80](https://github.com/wordpress-mobile/WordPress-Android/pull/18013/commits/b79fa80220b88a4842794bafede5a4a32aefe075)

**Pull request merged**
